python/ray/serve/tests/test_backend_state.py

```diff
@@ -1,89 +1,258 @@
-    backend_info = BackendInfo(
-        worker_class=lambda x: x,
-        backend_config=BackendConfig(),
-        replica_config=ReplicaConfig(lambda x: x))
-    if num_replicas:
-        backend_info.backend_config.num_replicas = num_replicas
+class MockTimer:
+    def __init__(self, start_time=None):
+        if start_time is None:
+            start_time = time.time()
+        self._curr = start_time
 
-    return backend_info
+    def time(self):
+        return self._curr
+
+    def advance(self, by):
+        self._curr += by
 
 
 @pytest.fixture
-def mock_backend_state_inputs() -> Tuple[BackendState, Mock, Mock]:
+def mock_backend_state() -> Tuple[BackendState, Mock, Mock]:
+    timer = MockTimer()
+    mock_replicas = []
     with patch(
-            "ray.serve.kv_store.RayInternalKVStore") as mock_kv_store, patch(
-                "ray.serve.long_poll.LongPollHost") as mock_long_poll, patch(
-                    "ray.serve.async_goal_manager.AsyncGoalManager"
-                ) as mock_goal_manager:
+            "ray.serve.backend_state.ActorReplicaWrapper",
+            new=mock_replica_factory(mock_replicas)), patch(
+                "time.time",
+                new=timer.time), patch("ray.serve.kv_store.RayInternalKVStore"
+                                       ) as mock_kv_store, patch(
+                                           "ray.serve.long_poll.LongPollHost"
+                                       ) as mock_long_poll, patch.object(
+                                           BackendState,
+                                           "_checkpoint") as mock_checkpoint:
         mock_kv_store.get = Mock(return_value=None)
+        goal_manager = AsyncGoalManager()
         backend_state = BackendState("name", True, mock_kv_store,
-                                     mock_long_poll, mock_goal_manager)
-        yield backend_state, mock_kv_store, mock_long_poll, mock_goal_manager
+                                     mock_long_poll, goal_manager)
+        mock_checkpoint.return_value = None
+        yield backend_state, timer, mock_replicas, goal_manager
+
+
+def replica_count(backend_state, backend=None, states=None):
+    total = 0
+    for backend_tag, state_dict in backend_state._replicas.items():
+        if backend is None or backend_tag == backend:
+            for state, replica_list in state_dict.items():
+                if states is None or state in states:
+                    total += len(replica_list)
+
+    return total
+
+
+def test_override_goals(mock_backend_state):
+    backend_state, _, _, goal_manager = mock_backend_state
+
+    b_config_1, r_config_1 = generate_configs()
+    initial_goal = backend_state.create_backend("tag1", b_config_1, r_config_1)
+    assert not goal_manager.check_complete(initial_goal)
+
+    b_config_2, r_config_2 = generate_configs(num_replicas=2)
+    new_goal = backend_state.create_backend("tag1", b_config_2, r_config_2)
+    assert goal_manager.check_complete(initial_goal)
```

Comment:
How is the initial goal complete now if `update()` was never called?  I'm comparing this to `test_create_delete_single_replica()` where `update()` needs to be called each time to advance the state.

Reply:
ah yeah it's sort of a weird behavior that `update()` needs to be called for the others. the difference here is that the original goal is completed immediately once an "overwriting" API calls happens, not when an async operation happens in the update loop (e.g., a replica is ready)